### PR TITLE
JIB Integration

### PR DESCRIPTION
**Pull request by theexplorist**

### CODE REVIEW

File: core/src/main/java/io/fabric8/maven/core/service/BuildService.java [hunk 1]
```diff
@@ -38,6 +43,8 @@
      */
     void build(ImageConfiguration imageConfig) throws Fabric8ServiceException;
 
+    void build(JibBuildConfigurationUtil buildConfigurationUtil) throws Fabric8ServiceException, InvalidImageReferenceException, InterruptedException, RegistryException, IOException, ExecutionException;
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/BuildService.java]: Interface method should only throw `Fabric8ServiceException`

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/DockerBuildService.java [hunk 2]
```diff
@@ -57,6 +61,11 @@ public void build(ImageConfiguration imageConfig) throws Fabric8ServiceException
         }
     }
 
+    @Override
+    public void build(JibBuildConfigurationUtil buildConfigurationUtil) throws Fabric8ServiceException, InvalidImageReferenceException, InterruptedException, RegistryException, IOException,  ExecutionException {
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/DockerBuildService.java]: Removes all exceptions and in the body of the method just throw `new UnsupportedOperationExcetion()`

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildService.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @Override
+    public void build(JibBuildConfigurationUtil buildConfigurationUtil) throws Fabric8ServiceException, InvalidImageReferenceException, InterruptedException, RegistryException, IOException, ExecutionException {
+       try {
+           jibBuild.buildImage(buildConfigurationUtil);
+       }  catch (Exception ex) {
+           throw new Fabric8ServiceException("Error while trying to build the image", ex);
+       }
+    }
+
+    public void build(ImageConfiguration imageConfiguration) throws Fabric8ServiceException {
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildService.java]: throw an `UnsupportedOperationException`

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    public JibBuildServiceUtil() {
+
+    }
+
+    public void buildImage(JibBuildConfigurationUtil buildConfigurationUtil) throws Fabric8ServiceException, InvalidImageReferenceException , InterruptedException, RegistryException, IOException,  ExecutionException {
+
+        String fromImage = buildConfigurationUtil.getFrom();
+        String targetImage = buildConfigurationUtil.getTo();
+        Map<String, String> credMap = buildConfigurationUtil.getCredMap();;
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: Removes extra semi colons.

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        return portSet;
+    }
+
+    protected void buildImage(String baseImage, String targetImage, Map<String, String> envMap, Map<String, String> credMap, Set<Port> portSet, List<Path> dependencyList) throws InvalidImageReferenceException , InterruptedException, RegistryException, IOException,  ExecutionException {
+
+        JibContainerBuilder contBuild = Jib.from(baseImage);
+
+        if(dependencyList.size() > 0) {
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: `!dependencyList.isEmpty()`

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            contBuild = contBuild.addLayer(dependencyList, AbsoluteUnixPath.get("/path/in/container"));
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: What is this path (`/path/in/container`)?

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        if(envMap.size() > 0) {
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: `use !isEmpty()`

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        if(credMap.size() > 0) {
+            String username = credMap.get("username");
+            String password = credMap.get("password");
+        }
+
+        try {
+            contBuild.containerize(
+                    Containerizer.to(RegistryImage.named(targetImage)
+                            .addCredential("myusername", "mypassword")));
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: why I should use `myusername`, `mypassword`? You should only call `addCredential` if there are credentials and it should be the configured credentials, not default ones (myusername, my password)

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        } catch (CacheDirectoryCreationException e) {
+            e.printStackTrace();
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: throw as IllegatStateException and not printing `stacktrace`

File: core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java [hunk 10]
```diff
@@ -144,6 +148,11 @@ public void build(ImageConfiguration imageConfig) throws Fabric8ServiceException
         }
     }
 
+    @Override
+    public void build(JibBuildConfigurationUtil buildConfigurationUtil) throws Fabric8ServiceException, InvalidImageReferenceException, InterruptedException, RegistryException, IOException, ExecutionException {
+
```
Comment (lordofthejars) [inline core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java]: throw the IllegalStateException

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java [hunk 11]
```diff
@@ -237,25 +259,68 @@ protected void executeInternal(ServiceHub hub) throws MojoExecutionException {
...
 
+    private void buildJibConfigUtil(List<ImageConfiguration> imageConfigList) {
+        for(ImageConfiguration imageConfig : imageConfigList) {
+            BuildImageConfiguration buildImgConfig = imageConfig.getBuildConfiguration();
+            JibConfigUtil configUtil = new JibConfigUtil(imageConfig);
+            JibBuildConfigurationUtil buildConfigurationUtil = configUtil.getUtil(buildImgConfig, authConfig, to, project);
+            try {
+                processImageConfig(hub, buildConfigurationUtil);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: Do not print stack traces, just throw an exception, check other mojos methods.

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java [hunk 12]
```diff
@@ -237,25 +259,68 @@ protected void executeInternal(ServiceHub hub) throws MojoExecutionException {
...
+            } catch (MojoExecutionException e) {
+                e.printStackTrace();
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: Do not print stack traces, just throw an exception, check other mojos methods.

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java [hunk 13]
```diff
@@ -310,6 +388,28 @@ protected void buildAndTag(ServiceHub hub, ImageConfiguration imageConfig)
                 .build();
     }
 
+    protected io.fabric8.maven.core.service.BuildService.BuildServiceConfig getBuildServiceConfigJib() throws MojoExecutionException {
+        return new io.fabric8.maven.core.service.BuildService.BuildServiceConfig.Builder()
```
Comment (lordofthejars) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: This is not Jib configuration so why the method is called as jib.

Comment (theexplorist) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: I am fetching the BuildServiceConfig in case we want to perform jib build

Comment (lordofthejars) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: But this is an openshift thing, and OpeShift works ith s2i to build the docker image.

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java (hunk 8 above)
Comment (theexplorist) [reply core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: Oh that is by mistake while checking then I pushed without changing it back!

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java (hunk 6 above)
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: @theexplorist we had a chat regarding this .. It can be `/app` . Make sure that your main class jar is also ADDed to /app path then.

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java (hunk 8 above)
Comment (dev-gaur) [reply core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: You have to use credentials from AuthConfig

Use the RegistryService class in Jib if you can.

File: core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
+package io.fabric8.maven.core.service.kubernetes;
+
+import com.google.cloud.tools.jib.api.*;
```
Comment (dev-gaur) [inline core/src/main/java/io/fabric8/maven/core/service/kubernetes/JibBuildServiceUtil.java]: Get rid of the wildcard imports

File: core/src/main/java/io/fabric8/maven/core/config/ProcessorConfig.java [hunk 15]
```diff
@@ -231,4 +231,4 @@ public static ProcessorConfig mergeProcessorConfigs(ProcessorConfig ... processo
         return ret.size() > 0 ? ret : null;
     }
 
-}
+}
```
Comment (dev-gaur) [inline core/src/main/java/io/fabric8/maven/core/config/ProcessorConfig.java]: Unncessary. remove this

File: generator/api/src/main/java/io/fabric8/maven/generator/api/support/BaseGenerator.java [hunk 16]
```diff
@@ -15,13 +15,19 @@
...
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 import io.fabric8.maven.core.config.OpenShiftBuildStrategy;
 import io.fabric8.maven.core.config.RuntimeMode;
+import io.fabric8.maven.core.model.Dependency;
+import io.fabric8.maven.core.model.GroupArtifactVersion;
```
Comment (dev-gaur) [inline generator/api/src/main/java/io/fabric8/maven/generator/api/support/BaseGenerator.java]: totally unnecessary changes in this file. remove.

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java [hunk 17]
```diff
@@ -64,8 +65,12 @@
     @Parameter
     private ProcessorConfig generator;
 
+    @Parameter(property = "fabric8.build.jib", defaultValue = "false")
+    private boolean isJib;
+
     /**
-     * Enrichers used for enricher build objects
+     * Enrichers used for enri @Parameter(property = "fabric8.to", defaultValue = "")
```
Comment (dev-gaur) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/BuildMojo.java]: Unnecessary.
